### ISSUE TITLE
Problem with nested quoting in environment variables

### ISSUE THREAD

**Issue by martin-lueders** [easybuilders/easybuild-framework#2845]
I am trying to modify a working easyconfig file by changing the toolchain from ("foss", "2018a") to ("fosscuda", "2018a"), leaving everything else the same.

The eb command now fails in the final (sanity checking) stage. I managed to track down the problem a bit further by locating the temporarily created module file. This contains the lines:
```
setenv  CUDA_CFLAGS             '-Xcompiler="-O2 -ftree-vectorize -march=native -fno-math-errno" -Xlinker="-L/home/octobld/eb-test/EasyBuild/software/CUDA/9.1.85-GCC-6.4.0-
2.28/lib64 -L/home/octobld/eb-test/EasyBuild/software/GCCcore/6.4.0/lib64 -L/home/octobld/eb-test/EasyBuild/software/GCCcore/6.4.0/lib -lm -lrt -lcudart -lpthread"'
```
and 
```
setenv  CUDA_CXXFLAGS           '-Xcompiler="-O2 -ftree-vectorize -march=native -fno-math-errno" -Xlinker="-L/home/octobld/eb-test/EasyBuild/software/CUDA/9.1.85-GCC-6.4.0-
2.28/lib64 -L/home/octobld/eb-test/EasyBuild/software/GCCcore/6.4.0/lib64 -L/home/octobld/eb-test/EasyBuild/software/GCCcore/6.4.0/lib -lm -lrt -lcudart -lpthread"'
```

These two lines give rise to an error
```
cmdSetenv.c(124):ERROR:11: Usage is 'setenv[-force] variable value'
bb-fosscuda-2018a(49):ERROR:102: Tcl command execution failed: setenv   CUDA_CFLAGS             '-Xcompiler="-O2 -ftree-vectorize -march=native -fno-math-errno" -Xlinker="-L/home/octobld/eb-test/EasyBuild/software/CUDA/9.1.85-GCC-6.4.0-2.28/lib64 -L/home/octobld/eb-test/EasyBuild/software/GCCcore/6.4.0/lib64 -L/home/octobld/eb-test/EasyBuild/software/GCCcore/6.4.0/lib -lm -lrt -lcudart -lpthread"'
```
when trying a module show or module load.

The similar looking line:
```
setenv           CUDA_CC 'nvcc -ccbin="g++"'
```
however works without problems. It seems related to the fact that the lines above contain more than one inner quotations.

Is there any easy way to fix this?

The easyconfig file I am using is:

```
easyblock = 'BuildEnv'

name = 'Octopus'
version = 'bb'

homepage = 'None'
description = """This module sets a group of environment variables for compilers, linkers, maths libraries, etc., required
 to compile Octopus."""

gccver = '6.4.0-2.28'

toolchain = {'name': 'fosscuda', 'version': '2018a'}

dependencies = [
    ('Autotools', '20170619')
]
                                            

moduleclass = 'devel'
```

**Comment by martin-lueders** [comment]
I did some further tests with the generated temporary module file.

If I swap (by hand) the single and double quotes, I can successfully load the module.  
The corresponding lines should look like:
```
setenv  CUDA_CXXFLAGS           "-Xcompiler='-O2 -ftree-vectorize -march=native -fno-math-errno' -Xlinker='-L/home/octobld/eb-test/EasyBuild/software/CUDA/9.1.85-GCC-6.4.0-
2.28/lib64 -L/home/octobld/eb-test/EasyBuild/software/GCCcore/6.4.0/lib64 -L/home/octobld/eb-test/EasyBuild/software/GCCcore/6.4.0/lib -lm -lrt -lcudart -lpthread'"
```
Unfortunately, I am not familiar enough with the internal workings of easybuild, so I do not know what to change. I hope that somebody more familiar can implement that.

Thanks,
Martin

**Comment by ocaisa** [comment]
@martin-lueders What version of EB are you using? I just want to be sure that this is still an issue with the latest release before looking deeper.

**Comment by martin-lueders** [comment]
@ocaisa 
eb --version:
This is EasyBuild 3.9.0 (framework: 3.9.0, easyblocks: 3.9.0)

**Comment by ocaisa** [comment]
I was googling this a bit and came across https://riptutorial.com/tcl/example/16957/quoting where it mentions you can use brackets to quote in Tcl. I tested swapping `'` for opening and closing brackets at the beginning and end and that seems to work. I think this would be the simpler solution in this case.

**Comment by ocaisa** [comment]
Can you confirm that that would instead and I'll do a bit of digging to see where we would need to make a change.

**Comment by martin-lueders** [comment]
Hi Alan,

Thanks for looking into this. The 'reversed' order of quotes seems to result in a module file which is loadable. I am not yet as far as to see whether I can build the software with it, as I have to change a whole chain of modules and eb files to include CUDA support. I will confirm as soon as I have a successful build.

Cheers,
Martin

**Comment by ocaisa** [comment]
I'd prefer not to go the route of reversing the quotations (I'd expect that could have a lot of unintended consequences), what I meant was can you check that
```
setenv  CUDA_CFLAGS             {-Xcompiler="-O2 -ftree-vectorize -march=native -fno-math-errno" -Xlinker="-L/home/octobld/eb-test/EasyBuild/software/CUDA/9.1.85-GCC-6.4.0-2.28/lib64 -L/home/octobld/eb-test/EasyBuild/software/GCCcore/6.4.0/lib64 -L/home/octobld/eb-test/EasyBuild/software/GCCcore/6.4.0/lib -lm -lrt -lcudart -lpthread"}
```
instead of
```
setenv  CUDA_CXXFLAGS           '-Xcompiler="-O2 -ftree-vectorize -march=native -fno-math-errno" -Xlinker="-L/home/octobld/eb-test/EasyBuild/software/CUDA/9.1.85-GCC-6.4.0-
2.28/lib64 -L/home/octobld/eb-test/EasyBuild/software/GCCcore/6.4.0/lib64 -L/home/octobld/eb-test/EasyBuild/software/GCCcore/6.4.0/lib -lm -lrt -lcudart -lpthread"'
```
works for you (it does for me). The only changes here are swapping the initial and final `'` for `{` and `}`

**Comment by ocaisa** [comment]
I think this would only require a change in https://github.com/easybuilders/easybuild-framework/blob/develop/easybuild/tools/utilities.py#L88 to
```
            return '{%s}' % val
```
for the tcl case.

@boegel There's already comment there that implies that fallback is *only* for Tcl, is that correct?

**Comment by ocaisa** [comment]
In the end in the PR #2847 , I've gone with just escaping the double quotation marks in the string.

**Comment by boegel** [comment]
@ocaisa I suspect the fallback comment you're referring to may date from the time before we even supported Lua syntax, so I wouldn't interpret it as "only need for Tcl, not for Lua"...

**Comment by martin-lueders** [comment]
Hi Alan,

still struggling with some other EB issues (fosscuda2018a and fosscuda2018b), but I tried the square brackets, the they don't seem to work. I just manually inserted them into the temporary module file, but this time, the file does not load. The reversed quotes from above do load.

Cheers,
Martin

**Comment by ocaisa** [comment]
For the record, they should be curly not square brackets but I didn't like the behaviour they had anyway (for example `$::env(HOME)` would not be evaluated) so I went with escaping the quotes instead.

Once I can get #2847 to pass tests, this problem should be solved.

**Comment by boegel** [comment]
Fixed in #2847, which will be included in the upcoming EasyBuild v3.9.1

Thanks a lot for reporting this issue @martin-lueders!
